File: app/src/main/java/ds/features/ui/RecyclerAdapter.java
package ds.features.ui;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import ds.features.R;
import ds.features.binding.WeatherViewModel;
import ds.features.databinding.ItemWeatherBinding;

import java.util.List;

public class RecyclerAdapter extends RecyclerView.Adapter<RecyclerAdapter.ViewHolder> {

	private final List<WeatherViewModel> items;
	private final Context ctx;
	private OnClickListener clickListener;


	public RecyclerAdapter(Context ctx, final List<WeatherViewModel> items) {
		this.ctx = ctx;
		this.items = items;

	}


	public void addItems(List<WeatherViewModel> newItems) {
		final int start = items.size();
		items.addAll(newItems);
		//notifyDataSetChanged();
		notifyItemRangeInserted(start, items.size());
	}


	@Override
	public ViewHolder onCreateViewHolder(final ViewGroup parent, final int position) {
		final ItemWeatherBinding binder = DataBindingUtil.inflate(LayoutInflater.from(ctx), R.layout.item_weather, parent, false);

		binder.setClicker(v -> clickListener.onClick(binder));
		return new ViewHolder(binder);
	}


	@Override
	public void onBindViewHolder(final ViewHolder viewHolder, final int position) {
		viewHolder.binder.setWeather(items.get(position));
	}


	@Override
	public int getItemCount() {
		return items.size();
	}


	public void setOnItemClickListener(OnClickListener l) {
		clickListener = l;

	}


	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
	public static class ViewHolder extends RecyclerView.ViewHolder {

		final ItemWeatherBinding binder;


		ViewHolder(ItemWeatherBinding binding) {
			super(binding.getRoot());
			this.binder = binding;
		}
	}


	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
	public interface OnClickListener {

		void onClick(ItemWeatherBinding binder);
	}
}

File: app/src/main/java/ds/features/ui/RecyclerAdapter.kt
package ds.features.ui

import android.content.Context
import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import ds.features.R
import ds.features.binding.WeatherViewModel
import ds.features.databinding.ItemWeatherBinding

class RecyclerAdapter(private val ctx: Context, private val items: MutableList<WeatherViewModel>) : RecyclerView.Adapter<RecyclerAdapter.ViewHolder>() {
	private var clickListener = fun(b: ItemWeatherBinding): Unit? = null


	fun addItems(newItems: List<WeatherViewModel>) {
		val start = items.size
		items.addAll(newItems)
		notifyItemRangeInserted(start, items.size)
	}


	override fun onCreateViewHolder(parent: ViewGroup, position: Int): ViewHolder {
		val binder = DataBindingUtil.inflate<ItemWeatherBinding>(LayoutInflater.from(ctx), R.layout.item_weather, parent, false)
		binder.setClicker { v -> clickListener(binder) }
		return ViewHolder(binder)
	}


	override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
		viewHolder.binder.weather = items[position]
		viewHolder.binder.executePendingBindings()
	}


	override fun getItemCount(): Int {
		return items.size
	}


	fun setOnItemClickListener(f: (ItemWeatherBinding) -> Unit) {
		clickListener = f
	}


	class ViewHolder(val binder: ItemWeatherBinding) : RecyclerView.ViewHolder(binder.getRoot())

}
